src/VisualStudio/Core/Impl/CodeModel/NodeKeyValidation.cs

```diff
@@ -17,14 +17,17 @@ public NodeKeyValidation()
 
         public void AddProject(AbstractProject project)
         {
-            var provider = (IProjectCodeModelProvider)project;
-            var fcms = provider.ProjectCodeModel.GetCachedFileCodeModelInstances();
-
-            foreach (var fcm in fcms)
+            var provider = project as IProjectCodeModelProvider;
+            if (provider != null)
```

Comment:
This is how we tie into Roslyn's workspace.  It provides the glue that maps between things like Documents and the IVsHierarchys and ItemIds that we *do* use.  Without this there is no correspondence between a random file in Solution explorer and a Roslyn document.  Without this there is no hookup between VS' editor fiters, and Roslyn's command handlers.  etc. etc.

"I though TypeScript didn't have a formal project system."

You are correct in that we don't have an *ms build* project system.  But we do have a project system.  It's just file system based.  i.e. we'll grovel your file system to determine the files that are in your 'virtual project'.  However, once we make this, we do need to both present a UI in solution explorer for this, and we need to hook into roslyn.  Using your roslyn project system bits makes this exceedingly simple.

By and large those project system bits are resilient and make no assumptions that the impl will be C# or VB.  But occasionally there is something like this that leaks through.